### ISSUE TITLE
`reportGeneralTypeIssues` in pytorch

### ISSUE THREAD

**Issue by lkct** [microsoft/pylance-release#4374]
<h3> Environment data </h3>

-   Language Server version: 2023.5.20
-   OS and version: linux x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.8.16 (miniconda)
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet </h3>

Install pytorch: `torch==2.0.1+cu117`
```python
import torch
torch.backends.cudnn.benchmark = False  # reportGeneralTypeIssues
```

##### Additional Context
In `torch/__init__.py` there's a line:
```python
import torch.backends.cudnn
```

<h3> Expected behavior </h3>

No error.

<h3> Actual behavior </h3>

"backends" is not a known member of module "torch" Pylance(reportGeneralTypeIssues)

**Comment by erictraut** [comment]
The `torch` library is a "py.typed" library, which means that it provides inlined type annotations. It also means that it follows certain rules about specifying which imported symbols it intends to re-export from its various modules. If you're interested, you can read about these rules [here](https://microsoft.github.io/pyright/#/typed-libraries?id=library-interface).

The top-level `torch` module doesn't re-export the `backends` submodule. It's worth noting that it does re-export several dozen other submodules including `fft`, `futures`, etc. So it appears that it was an explicit decision by the pytorch maintainers not to re-export `backends`. That means you need to import it explicitly in your code if you want to access it.

```python
import torch.backends
```

**Comment by lkct** [comment]
@erictraut Thanks very much for the explanation! I'll ask the pytorch community.

By the way, I'm curious if it's possible to re-export only `torch.backends.cudnn` in the top-level of `torch` without `torch.backends`. Could you please explain a little more?

**Comment by erictraut** [comment]
> I'm curious if it's possible to re-export only `torch.backends.cudnn` in the top-level of torch

It's possible to export the `torch.backends.cudnn` module and then re-export it as `cudnn` from the top-level module. Is that what you have in mind? To do so, the top-level module could use the statement `from torch.backends import cudnn as cudnn` or `import torch.backends.cudnn as cudnn`. Then your user code could simply `import torch` and use `torch.cudnn` to access that submodule.

It's also possible for the top-level module to import and re-export the `backends` submodule (`from torch import backends as backends`), and then the `backends` submodule could import and re-export the `cudnn` submodule from _its_ namespace (`from . import cudnn as cudnn`). If done this way, your user code could `import torch` and then use `torch.backends.cudnn` to access that submodule.

It all depends on how the torch maintainers want to define the library's public interface.

**Comment by lkct** [comment]
The first changes the interface, and I'm afraid they would not adopt it.

So is the second approach the only one to expose `torch.backends.cudnn` as is to users?

**Comment by erictraut** [comment]
Or you can import the submodule explicitly in your code: `import torch.backends.cudnn`. That's the only option that will work with the library as it is currently authored.

**Comment by lkct** [comment]
Thanks!

**Comment by malfet** [comment]
> The `torch` library is a "py.typed" library, which means that it provides inlined type annotations. It also means that it follows certain rules about specifying which imported symbols it intends to re-export from its various modules. If you're interested, you can read about these rules [here](https://microsoft.github.io/pyright/#/typed-libraries?id=library-interface).
> 
> The top-level `torch` module doesn't re-export the `backends` submodule. It's worth noting that it does re-export several dozen other submodules including `fft`, `futures`, etc. 

@erictraut is there any runtime check one can use to check if something is re-exported or not from `pyright` PoV?  I.e. both `fft` and `backends` are module members of torch, per:
```
% python -c "import inspect; import torch;print('backends' in sorted(x[0] for x in inspect.getmembers(torch, inspect.ismodule)))"
True
```

**Comment by erictraut** [comment]
> is there any runtime check one can use...?

Not that I can think of. You could perhaps walk the AST for the parent module to look for nodes that use redundant import forms. That would work for some cases, but it wouldn't account for cases where a type stub is present.

Out of curiosity, why do you want to verify this at runtime? Are you a library maintainer and want to validate this in a test? Or are you a library consumer?

**Comment by malfet** [comment]
> Not that I can think of. You could perhaps walk the AST for the parent module to look for nodes that use redundant import forms. That would work for some cases, but it wouldn't account for cases where a type stub is present.

Is there a PEP describing why it's proper to expose it one way vs the another?
 
> Out of curiosity, why do you want to verify this at runtime?

Because anything that is not covered by CI has a high probability of regressing again. 

>  Are you a library maintainer and want to validate this in a test? Or are you a library consumer?

Former.

**Comment by erictraut** [comment]
I've addressed the issue that caused the hang when running `pyright --verifytypes torch`. This fix will be included in version 1.1.310 of pyright.
